### PR TITLE
Add prometheus mode for kvdb tool

### PR DESCRIPTION
**Pull request by vstakhov**
This PR adds `prometheus` mode for kvdb, allowing to monitor database usage from Prometheus:

```
curl 'http://localhost:65432/metrics'
# HELP introspector_kvdb_elements_count Number of keys in kvdb
# TYPE introspector_kvdb_elements_count gauge
introspector_kvdb_elements_count{column="col0"} 1173995
introspector_kvdb_elements_count{column="col1"} 389466
introspector_kvdb_elements_count{column="col2"} 79
introspector_kvdb_elements_count{column="col3"} 15001
introspector_kvdb_elements_count{column="col4"} 10991
# HELP introspector_kvdb_keys_size Size of keys in KVDB
# TYPE introspector_kvdb_keys_size gauge
introspector_kvdb_keys_size{column="col0"} 48133795
introspector_kvdb_keys_size{column="col1"} 17334135
introspector_kvdb_keys_size{column="col2"} 3490
introspector_kvdb_keys_size{column="col3"} 330117
introspector_kvdb_keys_size{column="col4"} 560478
# HELP introspector_kvdb_values_size Size of keys in KVDB
# TYPE introspector_kvdb_values_size gauge
introspector_kvdb_values_size{column="col0"} 1445411701
introspector_kvdb_values_size{column="col1"} 9155436
introspector_kvdb_values_size{column="col2"} 125782
introspector_kvdb_values_size{column="col3"} 633760
introspector_kvdb_values_size{column="col4"} 52112606
```

### CODE REVIEW

File: src/kvdb/mod.rs [hunk 1]
```diff
@@ -324,6 +329,14 @@ fn run_with_db<D: IntrospectorKvdb>(db: D, opts: KvdbOptions) -> Result<()> {
 				KvdbDumpMode::Json => dump_into_json(db, dump_opts, output_dir.as_path())?,
 			};
 		},
+		KvdbMode::Prometheus(prometheus_opts) => {
+			match prometheus::run_prometheus_endpoint_with_db(db, prometheus_opts).await {
+				Ok(futures) => {
+					future::try_join_all(futures).await.map_err(|e| eyre!("Join error: {:?}", e))?;
+				},
+				Err(err) => error!("FATAL: cannot start prometheus kvdb command: {}", err),
```
Comment (sandreim) [inline src/kvdb/mod.rs:337]:
```suggestion
				Err(err) => error!("FATAL: cannot start kvdb command in prometheus mode: {}", err),
```

File: src/kvdb/mod.rs [hunk 2]
```diff
@@ -113,6 +115,8 @@ pub(crate) enum KvdbMode {
 	DecodeKeys(KvdbKeysOpts),
 	/// Dump database (works with a live database for RocksDB)
 	Dump(KvdbDumpOpts),
+	/// Work as prometheus endpoint providing kvdb metrics
```
Comment (sandreim) [inline src/kvdb/mod.rs:118]:
```suggestion
	/// Same as Usage, exposing metrics via a Prometheus endpoint
```

File: src/kvdb/prometheus.rs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+				.set(res.values_space);
+		}
+		info!("Ended update db iteration");
+		tokio::time::sleep(std::time::Duration::from_secs_f32(prometheus_opts.poll_timeout)).await;
+	}
+}
+
+fn register_metrics(registry: &Registry) -> KvdbPrometheusMetrics {
+	let elements_count_gauge = prometheus_endpoint::register(
+		IntGaugeVec::new(Opts::new("kvdb_elements_count", "Number of keys in kvdb"), &["column"]).unwrap(),
```
Comment (sandreim) [inline src/kvdb/prometheus.rs:119]: We could use expect for more explicit errors.

File: src/kvdb/prometheus.rs [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+async fn update_db<D: IntrospectorKvdb>(
+	db: D,
+	metrics: Arc<Mutex<KvdbPrometheusMetrics>>,
+	prometheus_opts: KvdbPrometheusOptions,
+) {
+	loop {
+		info!("Starting update db iteration");
+		let columns = db.list_columns().unwrap();
```
Comment (sandreim) [inline src/kvdb/prometheus.rs:78]: We could log an error here and not exit.

File: src/kvdb/prometheus.rs [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		let mut update_results: Vec<UpdateResult> = vec![];
+		for col in columns {
+			let mut keys_space = 0_i64;
+			let mut keys_count = 0_i64;
+			let mut values_space = 0_i64;
+
+			info!("Iterating over column {}", col.as_str());
+			let iter = db.iter_values(col.as_str()).unwrap();
```
Comment (sandreim) [inline src/kvdb/prometheus.rs:86]: Same here

File: src/kvdb/prometheus.rs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+	values_size_gauge: IntGaugeVec,
+	elements_count_gauge: IntGaugeVec,
+}
+
+pub async fn run_prometheus_endpoint_with_db<D: IntrospectorKvdb + Send + Sync + 'static>(
+	db: D,
+	prometheus_opts: KvdbPrometheusOptions,
+) -> Result<Vec<tokio::task::JoinHandle<()>>> {
+	let prometheus_registry = Registry::new_custom(Some("introspector".into()), None)?;
+	let metrics = Arc::new(Mutex::new(register_metrics(&prometheus_registry)));
```
Comment (sandreim) [inline src/kvdb/prometheus.rs:47]: Do we really need the mutex here?

**Review by sandreim — APPROVED** [review]

File: src/kvdb/prometheus.rs (hunk 5 above)
Comment (vstakhov) [reply src/kvdb/prometheus.rs:86]: That's a good point, as db can be indeed temporarily unavailable. However, as this db is not reopened, there is no point in not exiting here I think.

File: src/kvdb/prometheus.rs (hunk 6 above)
Comment (vstakhov) [reply src/kvdb/prometheus.rs:47]: No, we even don't need an arc, as gauges use atomic variables internally.

File: src/kvdb/prometheus.rs (hunk 3 above)
Comment (vstakhov) [reply src/kvdb/prometheus.rs:138]: I doubt that is useful as I cannot even imagine conditions why this could fail :)

**Pull request merged**